File: Kattis_python/1-D_frogger.py
numbers = input().split(' ')
n = int(numbers[0])
s = int(numbers[1])
m = int(numbers[2])
board = input().split(' ')

count = 0
game = True
visited = {}
while game:
    if s < 1:
        print("left")
        game = False
    elif s > n:
        print("right")
        game = False
    elif s in visited:
        print("cycle")
        game = False
    elif int(board[s-1]) == m:
        print("magic")
        game = False
    else:
        count += 1
        visited[s] = True
        s += int(board[s-1])
print(count)

File: Kattis/cpp/1-D_Frogger.cpp
#include <iostream>

int main()
{
    int n, s, m;
    std::cin >> n >> s >> m;
    // n is the number of squares on the board
    // s is the starting square using 1-index   (we'll fix that)
    // m is the target value that we want to reach
    s--;
    int *board = new int[n];
    int *visited_board = new int[n];

    for (int i = 0; i < n; i++)
    {
        std::cin >> board[i];
        visited_board[i] = 0;
    }

    int count = 0;

    while (true)
    {
        if (s >= n)
        {
            std::cout << "right" << std::endl;
            break;
        }
        else if (s < 0)
        {
            std::cout << "left" << std::endl;
            break;
        }
        else if (board[s] == m)
        {
            // this means we hit our magic number
            std::cout << "magic" << std::endl;
            break;
        }
        else if (visited_board[s] == 1)
        {
            std::cout << "cycle" << std::endl;
            break;
        }

        visited_board[s] = 1;
        s += board[s];
        count++;
    }
    std::cout << count << std::endl;

    delete[] board;
    delete[] visited_board;

    return 0;
}